docs/type-lists.md

```diff
@@ -0,0 +1,73 @@
+# Type Lists
+
+`T...` means a type list called `T...`. In a generic, if it is given a constraint, the constraint applies to each type in the list, not the list itself. So `[T...: U]` means a type list of any length such that each element is a subtype of `U`. The type `()` is an empty type list. Unlike a singleton type, this is not a subtype of the empty interface.
+
+`A...` is a tuple of zero or more types. `(A, B...)` is a tuple of one or more types. That is, the type list is flattened into the tuple. `A...->B` is a function type with zero or more arguments. `(A, B...)->C` is a function type with one or more arguments.
+
+`A...-->B` is a curried function type (note the use of `-->` instead of `->`). That is, the type list is flattened into the arrow. If `A...` is bound to `(C, D)`, you would get `C->D->B`.
+
+Parameters are tuples that may contain type lists. Calling a function does a destructuring bind of the tuple argument. Results are tuples that may contain type lists. Using the result also does a destructuring bind.
+
+```ts
+papply[A, B..., C, R: B->C](f: R~>(A, B...)->C, x: R~>A): R
+{
+  { y: B... => f(x, y) }
```

Comment:
May be important to indicate that tuple flattening also happens at the value level as well (i.e., this is not `(x, (y))`)?  And that this flattening is type-directed, based on the fact that `y : B...`.  If instead `z : (B...)` or `z : T` and `T ~ (B...)`, then `(x, z)` and `(x, (z))` are identical?

Reply:
Thanks, I've lifted you language!